### DISCUSSION TITLE
AWS EKS Add-ons Datasource

### DISCUSSION THREAD

**Discussion by adrianmace** [renovatebot/renovate#29465] [Suggest an Idea]
### Tell us more.

I propose a new datasource, similar to [aws-machine-image](https://docs.renovatebot.com/modules/datasource/aws-machine-image/), that queries AWS's `DescribeAddonVersions` [API](https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeAddonVersions.html).

Similar to the `aws-machine-image` datasource, I suspect it would be used like this:

```terraform
resource "aws_eks_addon" "vpc-cni" {
   cluster_name  = var.cluster_name
   addon_name    = "vpc-cni"
  # addonFilter=[{"Name":"addonName","Values":["vpc-cni"]},{"Name":"kubernetesVersion","Values":["1.29"]}]
   addon_version = "v1.18.1-eksbuild.3"
 }
```

Reference Issue: #23410

**Comment by secustor** [comment]
Have you tried to implement this using customDatasources?

**Reply by adrianmace** [reply]
I haven’t tried, but as the final comment in the related issue explains the API requires AWS authentication so there’s a bit of complexity involved which may warrant its own native one.

**Comment by jwilkicki** [comment]
I wrote a separate command that generates a custom datasource JSON file that has the versions for a given addon along with a regex manager for finding the versions for my code.  The problem I'm having now is RenovateBot picked up the  `eksbuild.x` part of the version string as a patch but is steadfastly refusing to update to any other version.  It's like it doesn't recognize that the components at the front (e.g. v1.18.1 in your example) are part of the version.  I'm not sure what to do about that.

**Reply by secustor** [reply]
Sounds like you are using a versioning which does not fit. 
If the last cipher is something like a build number / 4 version number. 
You probably need the regex and define that pattern.

**Reply by jwilkicki** [reply]
Tried semver and semver-coerced.  The format is exactly like in the example from the original poster.  I'll give the regex one a shot.

**Reply by secustor** [reply]
semver have to be exact SemVer 2.0 and `semver-coerced` simply translates version such as `1.2` to `1.2.0`. 
What you want to use is a said `regex` ( if you prefer precision ) or `loose` if you feel more lucky.

**Reply by taraspos** [reply]
@jwilkicki I think you are missing the `"ignoreUnstable": false`. I had the same problem.

https://docs.renovatebot.com/configuration-options/#ignoreunstable 

See my solution below:
- https://github.com/renovatebot/renovate/discussions/29465#discussioncomment-9959691

**Comment by jwilkicki** [comment]
Thanks to @secustor for pointing me in the right direction.  I can't share everything I did, but I'll try to share what I can for getting addon versions to work with custom datasources.

First, I wrote a shell script that calls the aws CLI and a couple of helpful CLI tools:  [semver-cli](https://github.com/davidrjonas/semver-cli) and [hcl2json](https://github.com/tmccombs/hcl2json)  The basic logic was to use `hcl2json` to convert my terraform file to json and then query the result with `jq` to get the current version we are using. I then used the aws CLI to call describeAddonVersion filtering on our Kubernetes version to get compatible versions.  

I then used semver-cli to parse our current version and the ones returned from AWS and filter down the list because AWS doesn't allow you to skip minor versions of some addons, so we couldn't have RenovateBot jump releases. We're just getting started with RenovateBot, so there might be a better way to do this with allowedVersions.

When I had the list of compatible versions with increments that we could jump from our current version, I used jq again to create a datasource.json file in the exact format that RenovateBot expects so that I didn't have to implement any JSONata filters to convert the results.

An example for coredns looks like this: 

```json
{
  "sourceUrl": "https://github.com/coredns/coredns",
  "homepage": "https://github.com/coredns/coredns",
  "releases": [
    {
      "version": "v1.9.3-eksbuild.15",
      "sourceUrl": "https://github.com/coredns/coredns"
    }
  ]
}
```

I have the script handle one addon at a time and create a separate JSON file for each datasource.  Once you have this, you could run it by hand or possibly run it on a schedule with your CI system to populate the file before your Renovate run.

The renovate.json part of this is going to be really dependent on what your code looks like, but here's a snippet that can point you in the right direction:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "baseBranches": [
    "main"
  ],
 "customManagers": [
    {
      "customType": "regex",
      "fileMatch": [
        "(^|/)some_terraform_file.tf"
      ],
      "matchStrings": [
        "create your regex here to match your version number in your HCL"
      ],
      "datasourceTemplate": "custom.{{depName}}",
      "versioningTemplate": "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-eksbuild.(?<build>\\d+)"
    }
  ],
  "customDatasources": {
    "vpc-cni": {
      "defaultRegistryUrlTemplate": "file://.github/vpc-cni-versions.json",
      "format": "json"
    },
    "coredns": {
      "defaultRegistryUrlTemplate": "file://.github/coredns-versions.json",
      "format": "json"
    },
    "kube-proxy": {
      "defaultRegistryUrlTemplate": "file://.github/kube-proxy-versions.json",
      "format": "json"
    },
    "aws-ebs-csi-driver": {
      "defaultRegistryUrlTemplate": "file://.github/aws-ebs-csi-driver-versions.json",
      "format": "json"
    }
  } 
}
```

**Comment by michemache** [comment]
Thanks, @adrianmace, for pushing this topic forward.
In the meantime, I use the following custom Datasource based on https://github.com/plumdog/eks-addon-configuration 

```
  "customDatasources": {
    "eks-addon": {
      "defaultRegistryUrlTemplate": "https://raw.githubusercontent.com/plumdog/eks-addon-configuration/main/data/{{packageName}}/addon.json",
      "format": "json",
      "transformTemplates": [
        "{ \"releases\": [ addonVersions[].{ \"version\": addonVersion } ] }",
      ],
    },
  },
  "packageRules": [
    {
      "matchDatasources": ["custom.eks-addon"],
      "versioning": "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-eksbuild\\.(?<build>\\d+)$",
      "commitMessageTopic": "EKS add-on {{depName}}",
    },
  ],
```

**Comment by taraspos** [comment]
I will share one more solution that works for self-hosted renovate or [renovatebot/github-action](https://github.com/renovatebot/github-action).

### Summary

We can fetch AWS EKS addons version via AWS CLI and transform it to the format needed by renovatebot to be used as Custom Datasource[^1]:

```bash
aws eks describe-addon-versions --region us-east-1 --kubernetes-version $EKS_VERISON \
        --query "addons[*].{addonName: addonName, releases: addonVersions[*].{version: addonVersion}}" --output json
```

<details><summary> JSON Output Format</summary>

```json
[
    {
        "addonName": "test-addon",
        "releases": [
            {
                "version": "v2.1.0-eksbuild.1"
            },
            {
                "version": "v1.1.0-eksbuild.1"
            },
            {
                "version": "v1.0.0-eksbuild.2"
            },
            {
                "version": "v1.0.0-eksbuild.1"
            }
        ]
    },
    {
        "addonName": "test-addon-2",
        "releases": [
            {
                "version": "v1.1.0-eksbuild.1"
            },
            {
                "version": "v1.0.0-eksbuild.1"
            }
        ]
    },
    {
        "addonName": "test-addon-3",
        "releases": [
            {
                "version": "v1.1.0"
            },
            {
                "version": "v1.0.0"
            }
        ]
    }
]
```

</details>

### Implementation details (Github Actions)

1. Configure custom datasource, custom manager and package group in renovate.json:

    ```json
    {
      /// ... renovate.json 
      "packageRules": [
        {
          "matchDatasources": ["custom.eks-addons"],
          "ignoreUnstable": false
        }
      ],
      "customDatasources": {
        "eks-addons": {
          "defaultRegistryUrlTemplate": "http://fileserver/versions/eks-addons.json",
          "transformTemplates": ["{'releases': $$[addonName = '{{packageName}}'].releases}"]
        }
      },
      "customManagers": [
        {
          "customType": "regex",
          "fileMatch": [
            "(^|\\/).*\\.tf$"
          ],
          "matchStrings": [
            "\\s?# renovate: datasource=eks-addons depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\s.*\\s?=\\s?\"(?<currentValue>.*)\",?\\s"
          ],
          "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}",
          "datasourceTemplate": "custom.eks-addons"
        }
      ]
      /// ... renovate.json 
    }
    ```

    - `ignoreUnstable`[^2] is needed to be disabled because of `-eksbuild.x` suffix in EKS addon versions
    - `defaultRegistryUrlTemplate`[^1] has to be an URL where we host file with versions, `file://` won't work because of https://github.com/renovatebot/renovate/discussions/29822#discussioncomment-9864079
    - `transformTemplates`[^1] is JSONata expression to lookup specific package from our object with eks-addon version, see https://try.jsonata.org/Jb1XhtyoQ
    - `customManagers`[^3] to match your need. the one above will work to lookup versions defined in terraform map

3. Configure github action that will tie it all together:
    1. Service container[^4] with Caddy webserver to host versions file:
       ```yaml
        services:
          fileserver:
            image: caddy:alpine
            volumes:
              - /tmp/versions/:/usr/share/caddy/versions/
            ports:
              - 80:80
       ```

    2. Steps to fetch EKS addon versions[^5] (requires valid AWS Credentials) and run Renovate itself:
        ```yaml
        steps:
        # All other steps you might need to get AWS credentials, checkout repo, etc
        # ...
        - name: Get EKS Addon versions
          run: |
              sudo chown -R $(id -u runner):$(id -g runner) /tmp/versions/
              aws eks describe-addon-versions --region us-east-1 --kubernetes-version $EKS_VERISON \
                --query "addons[*].{addonName: addonName, releases: addonVersions[*].{version: addonVersion}}" > /tmp/versions/eks-addons.json
        - name: Renovate
          uses: renovatebot/github-action:v40.2.0
          # ....
          with:
            # docker-network to be able to communicate with `http://fileserver/versions/eks-addons.json
            docker-network: ${{ job.container.network }}
         ```





----

[^1]: https://docs.renovatebot.com/modules/datasource/custom/
[^2]: https://docs.renovatebot.com/configuration-options/#ignoreunstable
[^3]: https://docs.renovatebot.com/configuration-options/#custommanagers
[^4]: https://docs.github.com/en/actions/using-containerized-services/about-service-containers
[^5]: https://docs.aws.amazon.com/cli/latest/reference/eks/describe-addon-versions.html

**Comment by ivankatliarchuk** [comment]
functionality released https://github.com/renovatebot/renovate/pull/33272
